lib/portage/util/file_copy.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+import errno
+import fcntl
+import logging
+import os
+import platform
+import shutil
+import sys
+
+
+logger = logging.getLogger(__name__)
+
+# Added in Python 3.12
+FICLONE = getattr(fcntl, "FICLONE", 0x40049409)
+
+# Unavailable in PyPy
+SEEK_DATA = getattr(os, "SEEK_DATA", 3)
+SEEK_HOLE = getattr(os, "SEEK_HOLE", 4)
+
+
+def _get_chunks(src):
+    try:
+        offset_hole = 0
+        while True:
+            try:
+                # Find the next bit of data
+                offset_data = os.lseek(src, offset_hole, SEEK_DATA)
+            except OSError as e:
+                # Re-raise for unexpected errno values
+                if e.errno not in (errno.EINVAL, errno.ENXIO):
+                    raise
+
+                offset_end = os.lseek(src, 0, os.SEEK_END)
+
+                if e.errno == errno.ENXIO:
+                    # End of file
+                    if offset_end > offset_hole:
+                        # Hole at end of file
+                        yield (offset_end, 0)
+                    break
+
+                # SEEK_DATA failed with EINVAL, return the whole file
+                yield (0, offset_end)
+                break
+            else:
+                offset_hole = os.lseek(src, offset_data, SEEK_HOLE)
+                yield (offset_data, offset_hole - offset_data)
+    except OSError:
+        logger.warning("_get_chunks failed unexpectedly", exc_info=sys.exc_info())
```

Comment:
```suggestion
    except OSError:
        logger.warning("_get_chunks failed unexpectedly", exc_info=sys.exc_info())
        raise
```

This needs to re-raise, or you won't copy the whole file.  This re-raise will trigger the `shutil.copyfile` fallback.

Reply:
Good catch.